plugins/slack_app/app/controllers/slack_app_controller.rb

```diff
@@ -0,0 +1,149 @@
+      SlackIdentifier.create!(user_id: current_user.id, identifier: resp['user_id'])
+
+      redirect_to
+    end
+  end
+
+  def command
+    check_slack_inputs! params
+    render json: command_response
+  end
+
+  def interact
+    # Interesting parts of the payload:
+    # {
+    #   "actions"=>[
+    #     {"name"=>"one", "value"=>"one"} <-- which button the user clicked
+    #   ],
+    #   "callback_id"=>"deploy-123456", <-- maps to a deployment
+    #   "user"=>{"id"=>"U0HAGH3AB", "name"=>"ben"}, <-- who clicked
+    #   "action_ts"=>"1468259216.482422",
+    #   "message_ts"=>"1468259019.000006", <-- need this to maybe update the message
+    #   "response_url"=>"https://hooks.slack.com/actions/T0HAGP0J2/58691408951/iHdjN2zjX0dZ5rXyl84Otql7"
+    # }
+    payload = JSON.parse(params[:payload])
+    check_slack_inputs!(payload)
+    render json: interact_response(payload) # This replaces the original message
+  end
+
+  def check_slack_inputs!(params)
+    if params[:ssl_check]
+      render json: 'ok' if params['ssl_check']
+      return false
+    end
+    raise "Invalid Slack token" unless params['token'] == ENV['SLACK_VERIFICATION_TOKEN']
+  end
+
+  def command_response
+    # Interesting parts of params:
+    # {
+    #   "user_id"=>"U0HAGH3AB", <-- look up Samson user using this
+    #   "command"=>"/deploy",
+    #   "text"=>"foo bar baz", <-- this is what to deploy
+    #   "response_url"=>"https://hooks.slack.com/commands/T0HAGP0J2/58604540277/g0xd4K2KOsgL9zXwR4kEc0eL"
+    #    ^^^^ This is how to respond later, we can also directly render some JSON
+    # }
+    deployer_identifier = SlackIdentifier.find_by_identifier params['user_id']
+    deployer = deployer_identifier.try(:user)
+    return unknown_user unless deployer
+
+    # Parse the command
+    projectname, branchname, stagename = params['text'].match(/([^\s\/]+)(?:\/(\S+))?\s*(?:to\s+(.*))?/).captures
+
+    # Sanity checks
+    begin
+      project = Project.find_by_param! projectname
+    rescue ActiveRecord::RecordNotFound
+      return unknown_project(projectname)
+    end
+
+    return unauthorized_deployer unless deployer.deployer_for?(project)
+
+    begin
+      stage = project.stages.find_by_param!(stagename || 'production')
+    rescue ActiveRecord::RecordNotFound
+      return unknown_stage(project, stagename)
+    end
+
+    deploy_service = DeployService.new(deployer)
+    deploy = deploy_service.deploy!(stage, reference: branchname || 'master')
+    DeployResponseUrl.create! deploy: deploy, response_url: params[:response_url]
+
+    SlackMessage.new(deploy).message_body
+  end
+
+  def interact_response(payload)
+    # This means someone clicked the "Approve" button in the channel
+    deploy = Deploy.find_by_id payload['callback_id']
+    return '_(Whoops, this deploy disappeared.)_' unless deploy
+
+    # Is this user connected?
+    buddy_identifier = SlackIdentifier.find_by_identifier payload['user']['id']
+    buddy = buddy_identifier.try(:user)
```

Comment:
It may be better to standardize getting the current slack user as a before_action and perform validation that user exists before getting into the main logic of handling the message.

It isn't too hard to figure out what's going on, but a lot of exit points in this that may be difficult to explain.